### PR TITLE
build: create api endpoint for accounts

### PR DESCRIPTION
**Pull request by mgnsharon**

### CODE REVIEW

File: apps/api/src/index.ts [hunk 1]
```diff
@@ -18,8 +16,10 @@ export default {
   async fetch(request: Request, env: APIEnv): Promise<Response> {
     const req = request.clone();
     const { url, method, headers } = request;
+
     const path = new URL(url).pathname;
-    if (method !== 'GET' && method !== 'OPTIONS') {
+
+    if (method !== 'GET' && method !== 'OPTIONS' && method !== 'POST') {
```
Comment (jeffywu) [inline apps/api/src/index.ts:22]: Doesn't this allow someone to POST to any endpoint?

File: packages/durable-objects/src/accounts-do.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      return new Response((error as Error).message, { status: 400 });
+    }
+  }
+
+  async update(request: Request) {
+    try {
+      const { events }: SentinalRequest = await request.json();
+      if (events && events.length > 0) {
+        const network = events[0].sentinel.network;
+        const accounts = unique(events.map((e) => e.transaction.from));
```
Comment (jeffywu) [inline packages/durable-objects/src/accounts-do.ts:38]: I don't think this is the right address, the SentinalRequest should have a variable called `account` on it. The address being stored here is the transaction sender (which may not be the actual address we care about).

File: packages/durable-objects/src/accounts-do.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        const existingAccounts =
+          (await this.state.storage.get<string[]>(network)) ?? [];
+        if (network && accounts) {
+          const newAccts = (accounts as string[]).filter(
+            (a: string) => !existingAccounts.includes(a)
+          );
+          existingAccounts.push(...newAccts);
+          await this.state.storage.put(network, existingAccounts);
+          return new Response('Accounts Updated', {
```
Comment (jeffywu) [inline packages/durable-objects/src/accounts-do.ts:47]: This should log to datadog as well?

File: packages/durable-objects/src/accounts-do.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  async fetch(request: Request): Promise<Response> {
+    try {
+      const req = request.clone();
+      const { method } = request;
+      if (method !== 'POST' && method !== 'GET') {
+        return new Response('Method Not Allowed', { status: 405 });
+      }
+      if (method === 'POST') {
```
Comment (jeffywu) [inline packages/durable-objects/src/accounts-do.ts:21]: You will probably need to authenticate the request here, right?

File: apps/api/src/index.ts (hunk 1 above)
Comment (mgnsharon) [reply apps/api/src/index.ts:22]: No, just the accounts endpoint, I'm looking at how I can lock it down

File: packages/durable-objects/src/accounts-do.ts (hunk 2 above)
Comment (mgnsharon) [reply packages/durable-objects/src/accounts-do.ts:38]: That's what the existing auto task was using, it recorded my wallet addresses.  I can double check there isn't another field we should use

File: packages/durable-objects/src/accounts-do.ts (hunk 4 above)
Comment (mgnsharon) [reply packages/durable-objects/src/accounts-do.ts:21]: I could create some kind of token that we could put in the url as a param possibly

Comment (jeffywu) [reply packages/durable-objects/src/accounts-do.ts:21]: We might also be able to check the `Host` header in the request in addition to the token as a second layer of protection. It seems like Host can be faked if Cloudflare chooses not to verify it, but I think you can test if Cloudflare does indeed verify it by trying to curl the endpoint from your local machine with a faked Host header.

Comment (mgnsharon) [reply packages/durable-objects/src/accounts-do.ts:21]: I looked at the web app firewall briefly, thought maybe we could allow posts only from the sentinel

**Review by jeffywu — APPROVED** [review]

**Pull request merged**